### PR TITLE
catalog: Remove upper fence error

### PR DESCRIPTION
**Pull request by jkosh44**
da7c2af16a3ad1a1a7454ea4ecc2eb55e645dc89 incorrectly asserted that the catalog debug tool does not fence out other writers when performing writes. The catalog debug tool does in face increment the epoch on every write, fencing out all other catalogs.

As a result, it's not possible for a process to write to the catalog without also fencing out all other catalog processes. Therefore, whenever a process encounters an upper mismatch error on write, there is always a more descriptive fencing error that could be returned.

This commit removes the `FenceError:Upper` variant and returns a more
descriptive error instead. Additionally, it adds some tests to validate
this behavior.

### Motivation
This PR fixes a previously unreported bug.

### Checklist

- [X] This PR has adequate test coverage / QA involvement has been duly considered. ([trigger-ci for additional test/nightly runs](https://trigger-ci.dev.materialize.com/))
- [X] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  
- [X] If this PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way), then it is tagged with a `T-proto` label.
- [X] If this PR will require changes to cloud orchestration or tests, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  
- [X] If this PR includes major [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note), I have pinged the relevant PM to schedule a changelog post.

### CODE REVIEW

**Comment by jkosh44** [comment]
> parallel-workload-materialized2-1    | thread 'tokio:work-5' panicked at src/catalog/src/durable/persist.rs:340:13: catalog write without fence encountered: UpperMismatch { expected: Antichain { elements: [447] }, current: Antichain { elements: [448] } }

lol, guess not. This should be impossible but clearly isn't.

**Comment by jkosh44** [comment]
> > parallel-workload-materialized2-1    | thread 'tokio:work-5' panicked at src/catalog/src/durable/persist.rs:340:13: catalog write without fence encountered: UpperMismatch { expected: Antichain { elements: [447] }, current: Antichain { elements: [448] } }
> 
> lol, guess not. This should be impossible but clearly isn't.

Ah OK, this can still happen when trying to write the new fence token itself.

**Comment by jkosh44** [comment]
I put this back in draft, hold off on reviewing it for a bit.

File: src/environmentd/src/deployment/preflight.rs [hunk 1]
```diff
@@ -215,22 +215,22 @@ pub async fn preflight_0dt(
 
                 match res {
                     Ok(_catalog) => break,
-                    Err(error) => {
-                        match error {
-                            CatalogError::Durable(
-                                mz_catalog::durable::DurableCatalogError::Fence(error),
```
Comment (jkosh44) [inline src/environmentd/src/deployment/preflight.rs:221]: It's still technically possible to get fenced while opening that catalog, but not by the old instance. We can only get fenced by a new instance that is trying to start up.

Comment (jkosh44) [reply src/environmentd/src/deployment/preflight.rs:221]: I have an idea for a follow up PR to move this retry logic into the catalog so we can remove this entirely. That should also speed up 0dt upgrades.

Comment (jkosh44) [reply src/environmentd/src/deployment/preflight.rs:221]: https://github.com/MaterializeInc/materialize/pull/29249

**Comment by jkosh44** [comment]
@aljoscha @ParkMyCar Ok, this is ready for review again.

File: src/catalog/tests/debug.rs [hunk 2]
```diff
@@ -252,3 +253,128 @@ async fn test_debug<'a>(state_builder: TestCatalogStateBuilder) {
...
+    assert_ok!(state.transaction().await);
+
+    // Edit catalog via debug_state.
+    debug_state
+        .edit::<SettingCollection>(
+            proto::SettingKey {
+                name: "joe".to_string(),
+            },
+            proto::SettingValue {
+                value: "koshakow".to_string(),
```
Comment (aljoscha) [inline src/catalog/tests/debug.rs:293]: 👀

File: src/environmentd/src/deployment/preflight.rs [hunk 3]
```diff
@@ -215,22 +215,22 @@ pub async fn preflight_0dt(
...
+                            error @ mz_catalog::durable::DurableCatalogError::UpperMismatch {
+                                ..
+                            },
+                        ) => {
+                            tracing::info!("upper mismatch while trying to fence out old deployment ({}), retrying...", error);
+                            continue;
                         }
-                    }
+                        error => {
+                            panic!(
```
Comment (aljoscha) [inline src/environmentd/src/deployment/preflight.rs:228]: This means we will panic when there's a new `envd` deployment/epoch/version, trying to do preflight checks/fencing? (Probably not a likely scenario, but ...)

**Review by aljoscha — APPROVED** [review]

File: src/environmentd/src/deployment/preflight.rs (hunk 3 above)
Comment (jkosh44) [reply src/environmentd/src/deployment/preflight.rs:228]: Yeah. We'll panic, reboot, then re-check the deployment version above to see if we should exit(0), halt, panic again, or take over. We could instead immediately re-check the deployment version and decide what to do here. Should I make that change?

**Review by ParkMyCar — APPROVED** [review]
LGTM from a catalog perspective!

**Pull request merged**